# Bagging - QRT ENS Data Challenge


## Libraries

In [ ]:
import numpy as np
import pandas as pd
from scipy.stats import spearmanr
import matplotlib.pyplot as plt
from sklearn.ensemble import BaggingRegressor
from sklearn.tree import DecisionTreeRegressor


## Loading data


In [ ]:
# After downloading the X_train/X_test/Y_train .csv files in your working directory:

X_train = pd.read_csv('X_train.csv')
y_train = pd.read_csv('y_train.csv')
X_test = pd.read_csv('X_test.csv')

In [ ]:
X_train.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1054     206      FR        0.210099       -0.427458       -0.606523   
1  2049     501      FR       -0.022399       -1.003452       -0.022063   
2  1924     687      FR        1.395035        1.978665        1.021305   
3   297     720      DE       -0.983324       -0.849198       -0.839586   
4  1101     818      FR        0.143807       -0.617038       -0.924990   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.606523            NaN       0.692860            NaN  ...   
1        0.022063      -0.573520      -1.130838       0.573520  ...   
2       -1.021305      -0.622021      -1.682587       0.622021  ...   
3        0.839586      -0.270870       0.563230       0.270870  ...   
4        0.924990            NaN       0.990324            NaN  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.444661 -0.172680 -0.556356 -0.790823 -0.283160 -1.069070   
1         -1.183194 -1.240300 -0.770457  1.522331  0.828412  0.437419   
2          1.947273 -0.480700 -0.313338  0.431134  0.487608  0.684884   
3         -0.976974 -1.114838 -0.507570 -0.499409 -0.236249  0.350938   
4         -0.526267 -0.541465 -0.424550 -1.088158 -1.011560  0.614338   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0 -0.063404  0.339041  0.124552   -0.002445  
1  1.831241 -0.659091  0.047114   -0.490365  
2  0.114836  0.535974  0.743338    0.204952  
3 -0.417514  0.911652 -0.296168    1.073948  
4  0.729495  0.245109  1.526606    2.614378  

[5 rows x 35 columns]

In [ ]:
y_train.head()

ID    TARGET
0  1054  0.028313
1  2049 -0.112516
2  1924 -0.180840
3   297 -0.260356
4  1101 -0.071733

## Bagging Regressor and train score

In [ ]:
# Create a decision tree regressor
X_train_clean = X_train.drop(['COUNTRY'], axis=1).fillna(0)
y_train_clean = y_train['TARGET']
dt = DecisionTreeRegressor()

# Create a bagging regressor using the decision tree as the base estimator
bagging_lr = BaggingRegressor(base_estimator=dt, n_estimators=10, random_state=42)

# Fit the bagging regressor on the data
bagging_lr.fit(X_train_clean, y_train_clean)

# Predict on the training set
output_train_bagging = bagging_lr.predict(X_train_clean)

# Define the metric function
def metric_train(output):
    return spearmanr(output, y_train_clean).correlation

# Calculate and print the Spearman correlation for the train set
print('Spearman correlation for the train set using bagging: {:.1f}%'.format(100 * metric_train(output_train_bagging)))


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Spearman correlation for the train set using bagging: 82.8%


The Spearman correlation obtained with our model on the train data set is about 82.8%.

## Generate the benchmark output

Next, we process the test set the same way as we did on the train set.

In [ ]:
X_test.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1115     241      FR        0.340083       -0.433604       -0.423521   
1  1202    1214      FR        0.803209        0.780411        0.601610   
2  1194    1047      FR        0.795540        0.721954        1.179158   
3  1084    1139      FR        0.172555       -0.723427       -0.044539   
4  1135     842      FR        0.949714        0.420236        0.617391   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.423521       0.165333       0.519419      -0.165333  ...   
1       -0.601610       0.342802       0.555367      -0.342802  ...   
2       -1.179158       1.620928       0.666901      -1.620928  ...   
3        0.044539            NaN      -0.205276            NaN  ...   
4       -0.617391       0.608561      -0.240856      -0.608561  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.222525 -0.513180 -0.182048 -0.982546 -0.876632  0.880491   
1          0.857739 -0.340595 -0.301094 -0.759816 -1.221443 -0.616617   
2          0.447967  0.796475 -0.367248  0.376055 -0.483363  0.865138   
3         -0.561295 -0.542606 -0.013291 -0.791119 -0.894309  0.239153   
4          0.503567 -0.230291 -0.609203 -0.744986 -1.196282  0.176557   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0  0.692242  0.569419 -0.029697   -0.929256  
1 -0.737496  0.251251  0.753646    0.664086  
2  0.120079 -1.485642 -0.326450   -0.349747  
3  0.457457 -0.746863  2.262654    0.642069  
4  0.312557 -2.219626 -0.509272   -0.488341  

[5 rows x 35 columns]

In [ ]:
X_test_clean = X_test.drop(['COUNTRY'], axis=1).fillna(0)

y_test_submission = X_test[['ID']].copy()
y_test_submission['TARGET'] = bagging_lr.predict(X_test_clean)

y_test_submission.to_csv('bagging.csv', index=False)


After submitting this bagging.csv file, we obtain a public score of around 7.39% indicating overfitting. In the next step, we will try to reduce overfitting by tuning hyperparameters

## Hyperparameter Tuning

In [ ]:
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import cross_val_predict
# Create the BaggingRegressor with DecisionTreeRegressor as the base estimator
bagging_lr = BaggingRegressor(base_estimator=DecisionTreeRegressor(), random_state=42)

# Define the parameter grid to search
param_grid = {
    'n_estimators': [100, 200, 300],
    'base_estimator__max_depth': [4, 5, 6],
    'max_samples': [0.5, 0.7, 0.8],
    'max_features': [0.5, 0.6, 0.7],
}

# Define the custom scoring function using Spearman correlation
def custom_spearman_scorer(estimator, X, y):
    predictions = estimator.predict(X)
    spearman_corr, _ = spearmanr(predictions, y)
    return spearman_corr

# Create the GridSearchCV object with the custom scoring function
grid_search = GridSearchCV(bagging_lr, param_grid, cv=5, scoring=custom_spearman_scorer)

# Fit the GridSearchCV object on the data
grid_search.fit(X_train_clean, y_train_clean)

# Get the best estimator from the grid search
best_bagging_lr = grid_search.best_estimator_

# Predict on the training set using the best model
output_train_best_bagging = best_bagging_lr.predict(X_train_clean)

# Print the best parameters and the Spearman correlation for the train set
print('Best Parameters:', grid_search.best_params_)
print('Spearman correlation for the train set using the best model: {:.1f}%'.format(100 * metric_train(output_train_best_bagging)))


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/en

Best Parameters: {'base_estimator__max_depth': 5, 'max_features': 0.5, 'max_samples': 0.7, 'n_estimators': 200}
Spearman correlation for the train set using the best model: 38.7%


After tuning our best set of parameters is 'base_estimator__max_depth': 5, 'max_features': 0.5, 'max_samples': 0.7, 'n_estimators': 200.
Spearman correlation for the train set using the best model is 38.7%.

In [ ]:
X_test_clean = X_test.drop(['COUNTRY'], axis=1).fillna(0)

y_test_submission = X_test[['ID']].copy()
y_test_submission['TARGET'] = best_bagging_lr.predict(X_test_clean)

y_test_submission.to_csv('bagging_tuned_final.csv', index=False)

After submitting the csv file on Challenge, we obtained a public score around 20.45% which is a significant imporvement from the default parameters.